libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -571,32 +585,48 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 			else
 				args.emplace_back(convert(*arguments[i], *parameterTypes[i]).commaSeparatedList());
 
-		if (auto identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+		FunctionDefinition const* functionDef;
```

Comment:
```suggestion
		FunctionDefinition const* functionDef = nullptr;
```

Reply:
I think that not having the default is better because then the compiler will warn you if you don't set it in one of the if branches. With a default you'll just get the call silently treated as a call via pointer, which will insert internal dispatch and will work but is not really the code we want to get.

Reply:
I'm not sure I want to rely on the compiler telling me. Mine doesn't for example.

Reply:
Also this is a class of bugs that usually does not appear in debug builds, which makes it even more dangerous.

Reply:
Oh, you're right. Turns out that clang warns about it but GCC does not. I'm usually compiling with GCC but I also use a syntax checker in the editor and it was issuing warnings because internally it runs clang. I totally agree that it might be hard to catch if compiler doesn't help us here.

Still, I in this case just assuming `nullptr` may have unintended consequences and those may not be easily noticed. I changed it to use `optional` so that I can detect it being uninitialized  with `solAssert(functionDef.has_value())`. Using `optional` with a pointer doesn't feel very elegant to me but pragmatically it works and is more robust...